algorithms/src/snark/varuna/ahp/ahp.rs

```diff
@@ -333,50 +326,44 @@ impl<F: PrimeField, MM: SNARKMode> AHPForR1CS<F, MM> {
-            let d_b = delta_b[i];
-            let d_c = delta_c[i];
-
-            let g_a_label = witness_label(id, "g_a", 0);
-            let g_a = LinearCombination::new(g_a_label.clone(), [(F::one(), g_a_label)]);
-            let g_a_at_gamma = evals.get_lc_eval(&g_a, challenges.gamma)?;
-            let g_b_label = witness_label(id, "g_b", 0);
-            let g_b = LinearCombination::new(g_b_label.clone(), [(F::one(), g_b_label)]);
-            let g_b_at_gamma = evals.get_lc_eval(&g_b, challenges.gamma)?;
-            let g_c_label = witness_label(id, "g_c", 0);
-            let g_c = LinearCombination::new(g_c_label.clone(), [(F::one(), g_c_label)]);
-            let g_c_at_gamma = evals.get_lc_eval(&g_c, challenges.gamma)?;
-
-            let sum_a = sums_fourth_msg[i].sum_a;
-            let sum_b = sums_fourth_msg[i].sum_b;
-            let sum_c = sums_fourth_msg[i].sum_c;
-
-            Self::add_g_m_term(&mut matrix_sumcheck, id, "a", challenges, g_a_at_gamma, v_rc, d_a, sum_a, s_a, RC)?;
-            Self::add_g_m_term(&mut matrix_sumcheck, id, "b", challenges, g_b_at_gamma, v_rc, d_b, sum_b, s_b, RC)?;
-            Self::add_g_m_term(&mut matrix_sumcheck, id, "c", challenges, g_c_at_gamma, v_rc, d_c, sum_c, s_c, RC)?;
-
-            linear_combinations.insert(g_a.label.clone(), g_a);
-            linear_combinations.insert(g_b.label.clone(), g_b);
-            linear_combinations.insert(g_c.label.clone(), g_c);
+            let RC = state_i.constraint_domain.size_as_field_element * state_i.variable_domain.size_as_field_element;
+
+            let matrices = ["a", "b", "c"];
+            let deltas = [delta_a[i], delta_b[i], delta_c[i]];
+            let non_zero_domains = [&state_i.non_zero_a_domain, &state_i.non_zero_b_domain, &state_i.non_zero_c_domain];
+            let sums = sums_fourth_msg[i].iter();
+
+            for (((m, sum), delta), non_zero_domain) in
+                matrices.into_iter().zip_eq(sums).zip_eq(deltas).zip_eq(non_zero_domains)
+            {
+                let selector_at_gamma = selector_evals(&mut selectors, &max_non_zero_domain, non_zero_domain, gamma);
+                let label = "g_".to_string() + m;
+                let g_m_label = witness_label(id, &label, 0);
+                let g_m = LinearCombination::new(g_m_label.clone(), [(F::one(), g_m_label)]);
+                let g_m_at_gamma = evals.get_lc_eval(&g_m, gamma)?;
+
+                let label_a_poly = format!("circuit_{id}_a_poly_{m}");
+                let a_poly = LinearCombination::new(label_a_poly.clone(), [(F::one(), label_a_poly)]);
+                let a_poly_missing = evals.get_lc_eval(&a_poly, gamma).is_err();
+                let label_b_poly = format!("circuit_{id}_b_poly_{m}");
+                let b_poly = LinearCombination::new(label_b_poly.clone(), [(F::one(), label_b_poly)]);
+                let b_poly_missing = evals.get_lc_eval(&b_poly, gamma).is_err();
+                assert_eq!(a_poly_missing, b_poly_missing);
```

Comment:
What logic is going on here? It could do with a comment, but why would `a_poly` and `b_poly` ever be missing at this point?

Reply:
Ah I see now. Maybe we can rewrite this to call a "evaluate_a_and_polys` method, which we implement differently for the prover and the verifier.

Reply:
Done 68f294cea

Note that I wanted to avoid passing a magic `prover: bool` to `construct_linear_combinations`, so instead I'm directly inferring whether a_poly and b_poly are present in `evals`.